README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->

# otpr <img src='man/figures/sticker.png' align="right" height=250/>

<!-- badges: start -->

[![Build Status](https://app.travis-ci.com/marcusyoung/otpr.svg?branch=master)](https://app.travis-ci.com/marcusyoung/otpr)
[![CRAN
status](https://www.r-pkg.org/badges/version/otpr)](https://cran.r-project.org/package=otpr)
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.4065250.svg)](https://doi.org/10.5281/zenodo.4065250)
[![Codecov test
coverage](https://codecov.io/gh/marcusyoung/otpr/branch/master/graph/badge.svg)](https://app.codecov.io/gh/marcusyoung/otpr?branch=master)
[![CRAN
downloads](https://cranlogs.r-pkg.org/badges/grand-total/otpr)](https://cran.r-project.org/package=otpr)
[![Project Status: Active – The project has reached a stable, usable
state and is being actively
developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
<!-- badges: end -->

## Overview

**otpr** is an R package that provides a wrapper for the
[OpenTripPlanner](https://www.opentripplanner.org/) (OTP) API. To use
**otpr** you will need a running instance of OTP. The purpose of the
package is to submit a query to the relevant OTP API resource, parse the
OTP response and return useful R objects. The package is aimed at both
new and expert users of OTP. The key parameters needed to query each
supported API resource are provided as (or via) **otpr** function
arguments. The argument values submitted by the user are comprehensively
checked prior to submission to OTP to ensure that they are valid and
make sense in combination, with feedback provided as appropriate. This
makes the package ideal for new users of OTP (especially when used with
the [accompanying
tutorial](https://github.com/marcusyoung/otp-tutorial)). Advanced users
can provide any additional API parameters they wish via the
`extra.params` argument. These parameters are passed directly to the OTP
API without checks.

**otpr** currently supports the following OTP API resources:

-   PlannerResource - retrieve one or more trip itineraries between
    supplied origin and destination by the designated mode(s).
-   LIsochrone Resource (OTPv1 only) - generate isochrone maps (the area
    accessible from or to a point within specified time thresholds).
-   SurfaceResource (OTPv1 only) - create and evaluate travel time
    surfaces (extremely efficient one-to-many analysis and generation of
    accessibility measures)

This package will be useful to researchers and transport planners who
want to use OTP to generate trip data for accessibility analysis or to
derive variables for use in transportation models.

## Version support

**otpr** fully supports OTP versions 1 and 2. Please note the following
:

-   There is a bug in OTPv2 which means that the first returned
    itinerary for TRANSIT trips is a (usually) sub-optimal WALK-only
    option. See:
    <https://github.com/opentripplanner/OpenTripPlanner/issues/3289>.
    Workaround: set an appropriate maxWalkDistance, perhaps 800
    (metres), but see below.  
-   In OTPv2 the `maxWalkDistance` parameter is treated as a hard limit
    when the mode is either WALK or (strangely) CAR or BICYCLE. This
    could result in no itinerary being returned if the parameter is
    change from its default (the default is NULL, which means that the
    OTP internal default of unlimited takes effect). This is different
    from the behaviour of OTPv1 where this parameter is effectively
    ignored when the mode is WALK and not applied at all to BICYCLE or
    CAR trips. See:
    <http://docs.opentripplanner.org/en/latest/OTP2-MigrationGuide/#rest-api>
    for more information.

## Installation

``` r
# Install from CRAN
# install.packages("otpr")
```

### Development version

To get a bug fix, or use a feature from the development version, you can
install otpr from GitHub. See
[NEWS](https://github.com/marcusyoung/otpr/blob/master/NEWS.md) for
changes since last release.

``` r
# install.packages("devtools")
# devtools::install_github("marcusyoung/otpr")
```

## Getting started

``` r
library(otpr)
```

### Defining an OTP connection

The first step is to call `otp_connect()`, which defines the parameters
needed to connect to a router on a running OTP instance. The function
can also confirm that the router is reachable.

``` r
# For a basic instance of OTP running on localhost with standard ports and a 'default' router
# this is all that's needed
otpcon <- otp_connect()
#> http://localhost:8080/otp is running OTPv1
#> Router http://localhost:8080/otp/routers/default exists
```

#### Handling Time Zones

If the time zone of an OTP graph *differs* from the time zone of the
local system running **otpr** then by default returned trip start and
end times will be expressed in the local system’s time zone and not the
time zone of the graph. This is because the OTP API returns EPOCH values
and the conversion to date and time format occurs on the local system. A
‘timeZone’ column is included in returned dataframes that contain start
and end times to make this explicit. If you wish to have start and end
times expressed in the time zone of the graph, the `tz` argument can be
specified when calling the `otp_connect()` function. This must be a
valid time zone (checked against the vector returned by `OlsonNames()`);
for example: “Europe/Berlin”.

## Querying the OTP API

### Function behaviour

The functions that query the OTP API return a list of three or more
elements. The first element is an errorId - with the value “OK” or the
error code returned by OTP. If errorId is “OK”, the second element will
contain the query response; otherwise it will contain the OTP error
message. There may be further list elements forming the query response.
The last element will be the query URL that was submitted to the OTP API
(for information and useful for troubleshooting).

### Distance between two points

To get the trip distance in metres between an origin and destination on
the street and/or path network use `otp_get_distance()`. You can specify
the required mode: CAR (default), BICYCLE or WALK are valid. The trip
information will relate to the *first* itinerary that was returned by
the OTP server.

``` r
# Distance between Manchester city centre and Manchester airport by CAR
otp_get_distance(
  otpcon,
  fromPlace = c(53.48805,-2.24258),
  toPlace = c(53.36484,-2.27108)
)
#> $errorId
#> [1] "OK"
#> 
#> $distance
#> [1] 29050.21
#> 
#> $query
#> [1] "http://localhost:8080/otp/routers/default/plan?fromPlace=53.48805,-2.24258&toPlace=53.36484,-2.27108&mode=CAR"

# Now for BICYCLE
otp_get_distance(
  otpcon,
  fromPlace = c(53.48805,-2.24258),
  toPlace = c(53.36484,-2.27108),
  mode = "BICYCLE"
)
#> $errorId
#> [1] "OK"
#> 
#> $distance
#> [1] 16028.57
#> 
#> $query
#> [1] "http://localhost:8080/otp/routers/default/plan?fromPlace=53.48805,-2.24258&toPlace=53.36484,-2.27108&mode=BICYCLE"
```

### Time between two points

To get the trip duration in minutes between an origin and destination
use `otp_get_times()`. You can specify the required mode: TRANSIT (all
available transit modes), BUS, RAIL, SUBWAY, TRAM, CAR, BICYCLE, and
WALK are valid. All the public transit modes automatically allow WALK.
There is also the option to combine TRANSIT with BICYCLE.

``` r
# Time between Manchester city centre and Manchester airport by BICYCLE
otp_get_times(
  otpcon,
  fromPlace = c(53.48805,-2.24258),
  toPlace = c(53.36484,-2.27108),
  mode = "BICYCLE"
)
#> $errorId
#> [1] "OK"
#> 
#> $duration
#> [1] 59.75
#> 
#> $query
#> [1] "http://localhost:8080/otp/routers/default/plan?fromPlace=53.48805,-2.24258&toPlace=53.36484,-2.27108&mode=BICYCLE&date=01-06-2022&time=19:07:48&walkReluctance=2&waitReluctance=1&arriveBy=FALSE&transferPenalty=0&minTransferTime=0"


# By default the date and time of travel is taken as the current system date and
# time. This can be changed using the 'date' and 'time' arguments
otp_get_times(
  otpcon,
  fromPlace = c(53.48805,-2.24258),
  toPlace = c(53.36484,-2.27108),
  mode = "TRANSIT",
  date = "01-19-2021",
  time = "07:15:00"
)
#> $errorId
#> [1] "OK"
#> 
#> $duration
#> [1] 39.65
#> 
#> $query
#> [1] "http://localhost:8080/otp/routers/default/plan?fromPlace=53.48805,-2.24258&toPlace=53.36484,-2.27108&mode=TRANSIT,WALK&date=01-19-2021&time=07:15:00&walkReluctance=2&waitReluctance=1&arriveBy=FALSE&transferPenalty=0&minTransferTime=0"
```

### Breakdown of time by mode, waiting time and transfers

To get more information about the trip when using transit modes,
`otp_get_times()` can be called with the `detail` argument set to TRUE.
The trip duration (minutes) is then further broken down by time on
transit, walking time (from/to and between stops), waiting time (when
changing transit vehicle or mode), and number of transfers (when
changing transit vehicle or mode). By default the function returns trip
information for the first itinerary suggested by the OTP server.
However, additional itineraries can be requested by specifying the
`maxItineraries` argument. The function will return every available
itinerary suggested by the OTP server, in order, *up to* the value of
`maxItineraries` (the default is 1).

``` r
# Time between Manchester city centre and Manchester airport by TRANSIT with detail
otp_get_times(
  otpcon,
  fromPlace = c(53.48805,-2.24258),
  toPlace = c(53.36484,-2.27108),
  mode = "TRANSIT",
  date = "01-19-2021",
  time = "07:15:00",
  detail = TRUE,
  maxItineraries = 1
)
#> $errorId
#> [1] "OK"
#> 
#> $itineraries
#>                 start                 end      timeZone duration walkTime
#> 1 2021-01-19 07:16:03 2021-01-19 07:55:42 Europe/London    39.65        8
#>   transitTime waitingTime transfers
#> 1          31        0.65         1
#> 
#> $query
#> [1] "http://localhost:8080/otp/routers/default/plan?fromPlace=53.48805,-2.24258&toPlace=53.36484,-2.27108&mode=TRANSIT,WALK&date=01-19-2021&time=07:15:00&walkReluctance=2&waitReluctance=1&arriveBy=FALSE&transferPenalty=0&minTransferTime=0"
```

#### Details of each leg for transit-based trips

To get information about each leg of transit-based trips,
`otp_get_times()` can be called with both the `detail` and `includeLegs`
arguments set to TRUE. A column called ‘legs’ will then be included in
the itineraries dataframe. This column contains a nested ‘legs’
dataframe for each itinerary. The ‘legs’ dataframe contains a row for
each leg of the trip. The information provided for each leg includes
start and end times, duration, distance, mode, route details, agency
details, and stop names. There is also a column called ‘departureWait’
which is the length of time in minutes required to wait before the start
of a leg. The sum of ‘departureWait’ will equal the total waiting time
for the itinerary.

``` r
# Time between Manchester city centre and Manchester airport by TRANSIT with detail and legs
trip <- otp_get_times(
  otpcon,
  fromPlace = c(53.48805,-2.24258),
  toPlace = c(53.36484,-2.27108),
  mode = "TRANSIT",
  date = "01-19-2021",
  time = "07:15:00",
  detail = TRUE,
  includeLegs = TRUE,
  maxItineraries = 2
)

# Legs for the first itinerary returned by OTP (first 9 columns)
trip$itineraries$legs[[1]][1:9]
#>             startTime             endTime      timeZone mode departureWait
#> 1 2021-01-19 07:16:03 2021-01-19 07:18:59 Europe/London WALK          0.00
#> 2 2021-01-19 07:19:00 2021-01-19 07:28:00 Europe/London TRAM          0.02
#> 3 2021-01-19 07:28:00 2021-01-19 07:29:23 Europe/London WALK          0.00
#> 4 2021-01-19 07:30:00 2021-01-19 07:52:00 Europe/London RAIL          0.62
#> 5 2021-01-19 07:52:01 2021-01-19 07:55:42 Europe/London WALK          0.02
#>   duration  distance routeType                 routeId
#> 1     2.93   197.370        NA                    <NA>
#> 2     9.00  1424.641         0 2:METLYELL:I:2021-01-02
#> 3     1.38   106.629        NA                    <NA>
#> 4    22.00 14921.306         2                  1:7285
#> 5     3.68   246.086        NA                    <NA>
```

### Travel time isochrones (OTPv1 only)

The `otp_get_isochrone()` function can be used to get one or more travel
time isochrones in either GeoJSON or SF format. These are only available
for transit or walking modes (OTP limitation). They can be generated
either *from* (default) or *to* the specified location.

#### GeoJSON example

``` r
# 900, 1800 and 2700 second isochrones for travel *to* Manchester Airport by any transit mode
my_isochrone <- otp_get_isochrone(
  otpcon,
  location = c(53.36484, -2.27108),
  fromLocation = FALSE,
  cutoffs = c(900, 1800, 2700),
  mode = "TRANSIT",
  date = "01-19-2021",
  time = "07:15:00"
)

# function returns a list of two elements
names(my_isochrone)
#> [1] "errorId"  "response" "query"

# now write the GeoJSON (in the "response" element) to a file so it can be opened in QGIS (for example)
write(my_isochrone$response, file = "my_isochrone.geojson")
```

#### SF example

``` r
# request format as "SF"
my_isochrone <- otp_get_isochrone(
  otpcon,
  location = c(53.36484, -2.27108),
  format = "SF",
  fromLocation = FALSE,
  cutoffs = c(900, 1800, 2700, 3600, 4500, 5400),
  mode = "TRANSIT",
  date = "01-19-2021",
  time= "07:15:00",
  maxWalkDistance = 1600,
  walkReluctance = 5,
  minTransferTime = 600
)

# plot using tmap package

library(tmap)
library(tmaptools)

# set bounding box
bbox <- bb(my_isochrone$response)
# get OSM tiles
osm_man <- read_osm(bbox, ext = 1.1)
# plot isochrones
tm_shape(osm_man) +
  tm_rgb() +
  tm_shape(my_isochrone$response) +
  tm_fill(
    col = "time",
    alpha = 0.8,
    palette = "-plasma",
    n = 6,
    style = "cat",
    title = "Time (seconds)"
  ) + tm_layout(legend.position = c("left", "top"), legend.bg.color = "white", 
                main.title = "15-90 minute isochrone to Manchester Airport", 
                main.title.size = 0.8)
```

<img src="man/figures/unnamed-chunk-9-1.png" width="80%" />

### One-to-many Travel Time Analysis (OTPv1 only)

If you wish to calculate the travel time from one or more origins to
many destinations, querying the OTP journey planning API (using
`otp_get_times()`) may not be the best option as it requires multiple
requests to the API which can be very inefficient. An alternative is to
use the OTP surface analysis feature, which enables you to calculate the
travel time from an origin to thousands of destinations in about the

NEWS.md
# otpr 0.5.1.9000 (Development)

# otpr 0.5.1

## Bug fixes

* Check that API call to evaluate surface actually returns some results. If not return error message - most likely cause is that the pointset coordinates are not within the area covered by the surface (e.g. could be the wrong pointset specified or latitude and longitude are the wrong way around).

# otpr 0.5.0

## New Features

* Support for creating and evaluating surfaces in OTPv1. The OTP surfaces endpoint enables
efficient running of one-to-many queries. Once a surface for an origin has been
generated (a surface is specific to a transport mode, date, time etc.), distances
from that origin to a set of locations (or 'opportunities') can be generated very
quickly - many thousands in a few seconds. In addition, one or more indicators of
accessibility for those opportunities can be generated at the same time. For example,
if the opportunity was census workplace zones, a count of the number of zones and the
number of jobs accessible from the origin at one-minute intervals are returned, along
with a cumulative sum, enabling a rapid assessment of the number of job opportunities
within, say, 30 minutes. There are two new functions to support surfaces:
    * `otp_create_surface()` creates a surface stored in memory on the OTP server and
    returns its Id number. It can optionally also retrieve and save a raster file (geoTIFF)
    of the surface for visualisation in R (or other GIS).
    * `otp_evaluate_surface()` uses an existing surface and a pointset (loaded from a CSV file
    provided to OTP at server startup) to calculate the travel time to each point location
    in the pointset file and to generate accessibility indicators for one or more 'opportunity'
    columns. A list is returned containing a dataframe for each 'opportunity' and, optionally,
    a dataframe of the time taken to travel from the surface's origin to each location point.
    This function can also be used to retrieve the time from an origin to many destinations.
* Support for OTPv2. Some issues to be aware of (due to OTPv2 implementation not **otpr**):
    * currently, due to a bug in OTPv2, for TRANSIT modes a sub-optimal walk-only trip may be returned
    as the top itinerary. See: [https://github.com/opentripplanner/OpenTripPlanner/issues/3289](https://github.com/opentripplanner/OpenTripPlanner/issues/3289) for a discussion of this issue. This can be mitigated by using the 
    `maxWalkDistance` parameter (but see below).
    * maxWalkDistance is a hard limit in OTPv2. This is a soft limit in OTPv1 and is effectively ignored
    if the mode is WALK only. In OTPv2 this parameter imposes a hard limit for WALK, CAR and BICYCLE modes. see: [http://docs.opentripplanner.org/en/latest/OTP2-MigrationGuide/#router-config](http://docs.opentripplanner.org/en/latest/OTP2-MigrationGuide/#router-config).
* Multiple trip itineraries are now supported by the `otp_get_times()` function and specified using the `MaxItineraries` argument.
* The `waitReluctance` argument has been added to `otp_get_times()`, `otp_get_isochrone()` and
`otp_get_surface()` functions.
* For advanced users: the ability to pass any *additional* OTP API parameter not specified
in the **otpr** functions to the OTP API via an `extra.params` argument. Available in
`otp_get_times()`, `otp_get_isochrone()` and `otp_get_surface()`. Note that no validation of
these additional parameters will be carried out by **otpr**. They will be passed directly to the OTP API.
* Now additionally imports **dplyr** and **rrapply**.

## Breaking changes

* As part of the support for multiple itineraries, trip legs are now returned by `otp_get_times()` as a nested dataframe within the
itineraries dataframe, rather than as a separate list element.
* The default value for maxWalkDistance (used in `otp_get_times()`, `otp_get_isochrone()` and `otp_create_surface()`) has been changed from 800 to NULL. This means that the internal OTP default of unlimited is applied. This has been implemented due to the impact of this parameter in OTPv2 (see above). It is now necessary to make a conscious decision on whether to set this parameter and to what value. The default value of 800 that was previously set by **otpr** was different from the OTP default (which is unlimited) and so this change corrects that anomaly.

## Bug fixes

* A single numeric value is now accepted for the `walkReluctance` parameter - this was incorrectly
restricted to an integer.

# otpr 0.4.2

## New Features

* `otp_get_distance()`, `otp_get_times()` and `otp_get_isochrone()` now also
return the query URL that was submitted to the OTP API. This is for user reference
and can be useful for troubleshooting.

* Now also imports **urltools**, used to return decoded query URL.

## Other

* Test coverage greatly extended.

# otpr 0.4.1

## Behaviour change

* `otp_get_times()` now returns a legs datafame, when requested, for WALK, BICYCLE 
and CAR modes. For these modes this will contain a single leg. This is for consistency
with other modes.

## Bug fixes

* Since 0.4.0 some parameters (`walkReluctance`, `arriveBy`, `transferPenalty` 
and `minTransferTime`) were not being passed to the OTP API by the
`otp_get_times()` function. This has been fixed.

* There were instances when `otp_get_times()` would generate an error when journey legs
were requested. This was due to an expected attribute in the OTP API response not 
being present. There is some inconsistency in the columns that are returned by OTP
for the legs, dependent on the mode, and what values appear in the original GTFS feeds.
Rather than limit the leg detail to the smaller subset of columns that are guaranteed
to be present, the function now includes columns in the `legs` dataframe for the attributes
that are present (as well as additional **otpr** calculated columns). This could mean some
inconsistency in the columns returned in the `legs` dataframe between queries.

R/otp_get_times.R
#' Queries OTP for the time or detailed itinerary for a trip between an origin
#' and destination
#'
#' In its simplest use case the function returns the time in minutes between an origin
#' and destination by the specified mode(s) for the top itinerary returned by OTP. If
#' \code{detail} is set to TRUE one or more detailed trip itineraries are returned,
#' including the time by each mode (if a multimodal trip), waiting time and the
#' number of transfers. Optionally, the details of each journey leg for each itinerary
#' can also be returned.
#'
#' @param otpcon An OTP connection object produced by \code{\link{otp_connect}}.
#' @param fromPlace Numeric vector, Latitude/Longitude pair, e.g. `c(53.48805, -2.24258)`
#' @param toPlace Numeric vector, Latitude/Longitude pair, e.g. `c(53.36484, -2.27108)`
#' @param mode Character vector, mode(s) of travel. Valid values are: WALK, BICYCLE,
#' CAR, TRANSIT, BUS, RAIL, TRAM, SUBWAY OR 'c("TRANSIT", "BICYCLE")'. TRANSIT will use all
#' available transit modes. Default is CAR. WALK mode is automatically
#' added for TRANSIT, BUS, RAIL, TRAM, and SUBWAY.
#' @param date Character, must be in the format mm-dd-yyyy. This is the desired date of travel.
#' Only relevant for transit modes. Default is the current system date.
#' @param time Character, must be in the format hh:mm:ss.
#' If \code{arriveBy} is FALSE (the default) this is the desired departure time, otherwise the
#' desired arrival time. Only relevant for transit modes. Default is the current system time.
#' @param arriveBy Logical. Whether a trip should depart (FALSE) or arrive (TRUE) at the specified
#' date and time. Default is FALSE.
#' @param maxWalkDistance Numeric. The maximum distance (in meters) that the user is
#' willing to walk. Default is NULL (the parameter is not passed to the API and the OTP
#' default of unlimited takes effect).
#' This is a soft limit in OTPv1 and is ignored if the mode is WALK only. In OTPv2
#' this parameter imposes a hard limit on WALK, CAR and BICYCLE modes (see:
#' \url{http://docs.opentripplanner.org/en/latest/OTP2-MigrationGuide/#router-config}).
#' @param walkReluctance A single numeric value. A multiplier for how bad walking is
#' compared to being in transit for equal lengths of time. Default = 2.
#' @param waitReluctance A single numeric value. A multiplier for how bad waiting for a
#' transit vehicle is compared to being on a transit vehicle. This should be greater
#' than 1 and less than \code{walkReluctance} (see API docs). Default = 1.
#' @param transferPenalty Integer. An additional penalty added to boardings after
#' the first. The value is in OTP's internal weight units, which are roughly equivalent to seconds.
#' Set this to a high value to discourage transfers. Default is 0.
#' @param minTransferTime Integer. The minimum time, in seconds, between successive
#' trips on different vehicles. This is designed to allow for imperfect schedule
#' adherence. This is a minimum; transfers over longer distances might use a longer time.
#' Default is 0.
#' @param detail Logical. When set to FALSE a single trip time is returned.
#' When set to TRUE one or more detailed trip itineraries are returned (dependent on \code{maxItineraries}).
#' Default is FALSE.
#' @param includeLegs Logical. Determines whether or not details of each
#' journey leg are returned. If TRUE then a nested dataframe of journeys legs will be returned
#' for each itinerary if \code{detail} is also TRUE. Default is FALSE.
#' @param maxItineraries Integer. Controls the number of trip itineraries that
#' are returned when \code{detail} is set to TRUE. This is not an OTP parameter.
#' All suggested itineraries are allowed to be returned by the OTP server. The function
#' will return them to the user in the order they were provided by OTP up to the maximum
#' specified by this parameter. Default is 1. This is an alternative to using the
#' OTP \code{maxNumItineraries} parameter which has problematic behaviour.
#' @param extra.params A list of any other parameters accepted by the OTP API PlannerResource entry point. For
#' advanced users. Be aware that otpr will carry out no validation of these additional
#' parameters. They will be passed directly to the API.
#' @return Returns a list of three or four elements. The first element in the list is \code{errorId}.
#' This is "OK" if OTP has not returned an error. Otherwise it is the OTP error code. The second element of list
#' varies:
#' \itemize{
#' \item If OTP has returned an error then \code{errorMessage} contains the OTP error message.
#' \item If there is no error and \code{detail} is FALSE then the \code{duration} in minutes is
#' returned as an integer. This is the duration of the top itinerary returned by the OTP server.
#'
#' \item If there is no error and \code{detail} is TRUE then \code{itineraries} as a dataframe.
#' }
#' The third element of the list is \code{query}. This is a character string containing the URL
#' that was submitted to the OTP API.
#' @details
#' If you plan to use the function in simple-mode - where just the duration of the top itinerary is returned -
#' it is advisable to first review several detailed itineraries to ensure that the parameters
#' you have set are producing sensible results.
#'
#' If requested using \code{includeLegs}, the itineraries dataframe will contain a column called 'legs'
#' which has a nested legs dataframe for each itinerary. Each legs dataframe will contain
#' a set of core columns that are consistent across all queries. However, as the OTP
#' API does not consistently return the same attributes for legs, there will be some variation
#' in columns returned. You should bare this in mind if your post processing
#' uses these columns (e.g. by checking for column existence).
#' @examples \dontrun{
#' otp_get_times(otpcon, fromPlace = c(53.48805, -2.24258), toPlace = c(53.36484, -2.27108))
#'
#' otp_get_times(otpcon, fromPlace = c(53.48805, -2.24258), toPlace = c(53.36484, -2.27108),
#' mode = "BUS", date = "03-26-2019", time = "08:00:00")
#'
#' otp_get_times(otpcon, fromPlace = c(53.48805, -2.24258), toPlace = c(53.36484, -2.27108),
#' mode = "BUS", date = "03-26-2019", time = "08:00:00", detail = TRUE)
#'}
#' @importFrom rlang .data
#' @importFrom dplyr any_of
#' @export
otp_get_times <-
  function(otpcon,
           fromPlace,
           toPlace,
           mode = "CAR",
           date = format(Sys.Date(), "%m-%d-%Y"),
           time = format(Sys.time(), "%H:%M:%S"),
           maxWalkDistance = NULL,
           walkReluctance = 2,
           waitReluctance = 1,
           arriveBy = FALSE,
           transferPenalty = 0,
           minTransferTime = 0,
           maxItineraries = 1,
           detail = FALSE,
           includeLegs = FALSE,
           extra.params = list())
  {
    # get the OTP parameters ready to pass to check function
    call <- sys.call()
    call[[1]] <- as.name('list')
    params <- eval.parent(call)
    params <-
      params[names(params) %in% c("mode", "detail", "includeLegs", "maxItineraries", "extra.params") == FALSE]
    
    # Check for required arguments
    if (missing(otpcon)) {
      stop("otpcon argument is required")
    } else if (missing(fromPlace)) {
      stop("fromPlace argument is required")
    } else if (missing(toPlace)) {
      stop("toPlace argument is required")
    }
    
    # function specific argument checks
    args.coll <- checkmate::makeAssertCollection()
    checkmate::assert_list(extra.params)
    checkmate::assert_logical(detail, add = args.coll)
    checkmate::assert_integerish(maxItineraries,
                                 lower = 1,
                                 add = args.coll)
    checkmate::reportAssertions(args.coll)
    
    # process mode
    mode <- otp_check_mode(mode)
    
    # OTP API parameter checks
    do.call(otp_check_params,
            params)
    
    # Construct URL
    routerUrl <- paste0(make_url(otpcon)$router, "/plan")
    
    # Collapse fromPlace and toPlace
    fromPlace <- paste(fromPlace, collapse = ",")
    toPlace <- paste(toPlace, collapse = ",")
    
    # Construct query list
    query <- list(
      fromPlace = fromPlace,
      toPlace = toPlace,
      mode = mode,
      date = date,
      time = time,
      maxWalkDistance = maxWalkDistance,
      walkReluctance = walkReluctance,
      waitReluctance = waitReluctance,
      arriveBy = arriveBy,
      transferPenalty = transferPenalty,
      minTransferTime = minTransferTime
    )

    # append extra.params to query if present
    if (length(extra.params) > 0) {
      msg <- paste("Unknown parameters were passed to the OTP API without checks:", paste(sapply(names(extra.params), paste), collapse=", "))
      warning(paste(msg), call. = FALSE)
      query <- append(query, extra.params)
    }
    
    # Use GET from the httr package to make API call and place in req - returns json by default.
    req <- httr::GET(routerUrl,
                     query = query)
    
    # decode URL for return
    url <- urltools::url_decode(req$url)
    
    # convert response content into text
    text <- httr::content(req, as = "text", encoding = "UTF-8")
    # parse text to json
    asjson <- jsonlite::fromJSON(text, flatten = TRUE)
    
    # Check for errors
    # Note that OTPv1 and OTPv2 use a different node name for the error message.
    if (!is.null(asjson$error$id)) {
      response <-
        list(
          "errorId" = asjson$error$id,
          "errorMessage" = ifelse(
            otpcon$version == 1,
            asjson$error$msg,
            asjson$error$message
          ),
          "query" = url
        )
      return (response)
    } else {
      error.id <- "OK"
    }
    
    # OTPv2 does not return an error when there is no itinerary - for
    # example if WALK mode and maxWalkDistance is too low. So now also check that
    # there is at least 1 itinerary present.
    if (length(asjson$plan$itineraries) == 0) {
      response <-
        list(
          "errorId" = -9999,
          "errorMessage" = "No itinerary returned. If using OTPv2 the maxWalkDistance parameter (default 800m) might be too restrictive. It 
          is applied by OTPv2 to BICYCLE and CAR modes in addition to WALK",
          "query" = url
        )
      return (response)
    }
    
    # check if we need to return detailed response
    if (detail == TRUE) {
      # Return up to maxItineraries
      num_itin <-
        pmin(maxItineraries, nrow(asjson$plan[["itineraries"]]))
      df <- asjson$plan$itineraries[c(1:num_itin),]
      # convert times from epoch format
      df$start <-
        otp_from_epoch(df$startTime, otpcon$tz)
      df$end <-
        otp_from_epoch(df$endTime, otpcon$tz)
      df$timeZone <- attributes(df$start)$tzone[1]
      # If legs are required we process the nested legs dataframes preserving
      # structure using rrapply
      if (isTRUE(includeLegs)) {
        # clean-up colnnames
        legs <-
          rrapply::rrapply(
            df$legs,
            f = function(x)
              janitor::clean_names(x, case = "lower_camel"),
            classes = "data.frame"
          )
        # convert from epoch times
        legs <-
          rrapply::rrapply(
            legs,
            condition = function(x, .xname)
              .xname %in% c("startTime", "endTime", "fromDeparture", "fromArrival"),
              f = function(x)
                otp_from_epoch(x, otpcon$tz)
          )
  
        
         # calculate departureWait - not relevant for one leg itineraries
        # (e.g. WALK only trip) where there won't be a fromArrival
        # However, for ease of processing of returned data we set departureWait
        # to zero for these.
        legs <-
          rrapply::rrapply(
            legs,
            f = function(x)
              if (nrow(x) > 1)
                dplyr::mutate(x, departureWait = round(abs((as.numeric(
                  .data$fromArrival - .data$fromDeparture
                )) / 60
                ), 2))
            else
              dplyr::mutate(x, departureWait = 0) ,
            classes = "data.frame"
          )
        # if departureWait is NA (usually for first leg of multi-leg trip) replace with 0
        legs <-
          rrapply::rrapply(
            legs,
            condition = function(x, .xname)
              .xname == "departureWait",
            f = function(x)    
            replace(x, is.na(x), 0)
          )
        # Update duration column to minutes
        legs <-
          rrapply::rrapply(
            legs,
            condition = function(x, .xname)
              .xname == "duration",
            f = function(x)  
              round(x / 60, 2)
          )
        # Add timezone column
        legs <-
          rrapply::rrapply(
            legs,
            f = function(x)
              dplyr::mutate(x, timeZone = attributes(x$startTime)$tzone[1]),
            classes = "data.frame"
          )
        # select required columns in legs using %in% as sometimes columns are missing
        # for example routeShortName. Also there are fewer columns when just a WALK,
        # BICYCLE or CAR leg is returned.
        leg_columns <- c(
          'startTime',
          'endTime',
          'timeZone',
          'mode',
          'departureWait',
          'duration',
          'distance',
          'routeType',
          'routeId',
          'routeShortName',
          'routeLongName',
          'headsign',
          'agencyName',
          'agencyUrl',
          'agencyId',
          'fromName',
          'fromLon',
          'fromLat',
          'fromStopId',
          'fromStopCode',
          'toName',
          'toLon',
          'toLat',
          'toStopId',
          'toStopCode'
        )
        # Select columns
        legs <-
          rrapply::rrapply(
            legs,
            f = function(x)
              dplyr::select(x, which(colnames(x) %in%
                                       leg_columns)),
            classes = "data.frame"
          )
        # change column order using relocate
        legs <- rrapply::rrapply(
          legs,
          f = function(x)
            dplyr::relocate(x, any_of(leg_columns)),
          classes = "data.frame"
        )
      } # end legs processing
      # subset the dataframe ready to return
      ret.df <-
        dplyr::select(
          df,
          c(
            'start',
            'end',
            'timeZone',
            'duration',
            'walkTime',
            'transitTime',
            'waitingTime',
            'transfers'
          )
        )
      # Insert processed legs if required
      if (isTRUE(includeLegs)) {
        ret.df$legs <- legs
      }
      # convert seconds into minutes where applicable
      ret.df[, 4:7] <- round(ret.df[, 4:7] / 60, digits = 2)
      # rename walkTime column as appropriate - this a mistake in OTP
      if (mode == "CAR") {
        names(ret.df)[names(ret.df) == 'walkTime'] <- 'driveTime'
      } else if (mode == "BICYCLE") {
        names(ret.df)[names(ret.df) == 'walkTime'] <- 'cycleTime'
      }
      response <-
        list("errorId" = error.id,
             "itineraries" = ret.df,
             "query" = url)
      return (response)
    } else {
      # detail not needed - just return travel time in minutes from the first itinerary
      response <-
        list(
          "errorId" = error.id,
          "duration" = round(asjson$plan$itineraries[1, "duration"] / 60, digits = 2),
          "query" = url
        )
      return (response)
    }
  }
